x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/slm/SnapshotRetentionTask.java

```diff
@@ -65,21 +54,18 @@
 public class SnapshotRetentionTask implements SchedulerEngine.Listener {
 
     private static final Logger logger = LogManager.getLogger(SnapshotRetentionTask.class);
-    private static final AtomicBoolean running = new AtomicBoolean(false);
```

Comment:
This flag seems unnecessary with concurrent snapshots. If (unlikely case now anyway) two delete runs collide and re-submit the same snapshot(s) for deletion the delete functionality will de-duplicate those deletes => we don't save anything by having this guard IMO and it's just more complexity (especially with the necessary changes in here making all the APIs async)

Reply:
One concern I have here is stats counting, if SLM retention is super aggressive and deletes the same snapshot 15 times (let's assume it's spending forever deleting the snapshot) then it tells the client that it's succeeded 15 times right? that means we'd increment the deleted stats incorrectly for the actual number of snapshots that have been deleted

Reply:
Thanks for spotting, yes this would've caused broken counting I think.
I fixed this in https://github.com/elastic/elasticsearch/pull/62284/commits/83dbef89b26f4242cee800a4fc1e13a9195e40da by keeping track of the currently deleting `SnapshotId`s. Admittedly there's still some non-zero chance of running into a race here if a delete just finished right after the loop over the snapshots to delete started but IMO it's ok to ignore that for now because it's unlikely, will only add to the failure counter and fixing it would way complicating the code. Also, it's maybe a nice pointer for people to reduce their retention frequency if they're seeing these kinds of issues?
(that said, if you think the race here might be an issue I'm happy to see what I can do about a fix, the main thing preventing a short and sweet fix here is the lack of access to the cluster service in this class which would allow fixing this by nicely ordering things on the CS thread like we do for other snapshot things)

That also helps prevent a bunch of listeners pilling up needlessly in this scenario.